thunder/dynamo/compiler.py

```diff
@@ -0,0 +1,62 @@
+import torch
+
+from collections import OrderedDict
+
+
+class ThunderCompiler:
+    def __init__(self, **thunder_options):
+        """
+        A class that compiles a `fx.GraphModule` to a `thunder.ThunderModule`.
+        This class is meant to be used as a backend for the `torch.compile`
+        function.
+
+        Keyword arguments:
+            thunder_options: a dictionary of options to pass to `thunder.jit`.
+
+        Example:
+            >>> import torch
+            >>> from thunder.dynamo import ThunderCompiler
+            >>> backend = ThunderCompiler()
+            >>> x = torch.ones(2, requires_grad=True)
+            >>> @torch.compile(backend=backend)
+            ... def func(x):
+            ...     x = torch.sin(x)
+            ...     if x.sum() > 0:
+            ...         return x + 1
+            ...     else:
+            ...         return x - 1
+            >>> out = func(x)
+        """
+        from thunder import ThunderModule
+
+        # Thunder-compiled functions should be readily available for inspection
+        # and testing, so we will store them in a list. The order of the
+        # functions in the list will be the same as the order in which they were
+        # compiled. In addition, we will store a mapping from the ThunderModule
+        # to the GraphModule that was passed to ThunderCompiler. This will allow
+        # us to inspect the GraphModule that was compiled by Thunder.
+        self.thunder_fns: list[ThunderModule] = []
+        self.thunder_to_gm: OrderedDict[ThunderModule, torch.fx.GraphModule] = OrderedDict()
+
+        self.thunder_options = thunder_options
+
+        # There will be pieces of Dynamo IR that Thunder cannot compile, so we
+        # will need to build a fallback mechanism to handle those cases.
+        # Possible stages of the compilation that need to be saved for inspection:
+        # 1. The GraphModule as it was passed to ThunderCompiler.
+        # 2. The GraphModule after split for Thunder/PyTorch.
+        # 3. If the whole GraphModule is not supported, record the reasons why.
+
+    def __call__(self, gm: torch.fx.GraphModule, sample_args: list[torch.SymInt, torch.Tensor]):
+        from thunder import jit
```

Comment:
Do we want to throw a warning (once) here that the backend is "experimental" or "in development"?

Reply:
Yes, let's do that!

Reply:
@apaz-cli has some functions to help warn once here:

https://github.com/Lightning-AI/lightning-thunder/blob/86cd9c3b5824064f8faa024ae0427fb297c04add/thunder/core/baseutils.py#L499

Reply:
I added a warning that is shown each time when ThunderCompiler is initialized.